buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -588,7 +588,7 @@ static ByteBuf encodeString0(ByteBufAllocator alloc, boolean enforceHeap, CharBu
             dst = alloc.buffer(length);
         }
         try {
-            final ByteBuffer dstBuf = dst.internalNioBuffer(0, length);
+            final ByteBuffer dstBuf = dst.internalNioBuffer(dst.readerIndex(), length);
```

Comment:
This behavior is not a strict requirement of the `ByteBufAllocator` interface (although it is reasonable behavior and our default implementations should do this).